hbase-client/src/main/java/org/apache/hadoop/hbase/client/RawAsyncTableImpl.java

```diff
@@ -796,75 +804,165 @@ private boolean locateFinished(RegionInfo region, byte[] endKey, boolean endKeyI
 
   private <S, R> void coprocessorServiceUntilComplete(Function<RpcChannel, S> stubMaker,
     ServiceCaller<S, R> callable, PartialResultCoprocessorCallback<S, R> callback,
-    AtomicBoolean locateFinished, AtomicInteger unfinishedRequest, RegionInfo region, Span span) {
+    MultiRegionCoprocessorServiceProgress<R> progress, RegionInfo region,
+    AtomicBoolean claimedRegionAssignment, Span span) {
     addListener(coprocessorService(stubMaker, callable, region, region.getStartKey()), (r, e) -> {
       try (Scope ignored = span.makeCurrent()) {
-        if (e != null) {
-          callback.onRegionError(region, e);
+        if (e instanceof RegionNoLongerExistsException) {
+          RegionInfo newRegion = ((RegionNoLongerExistsException) e).getNewRegionInfo();
+          if (progress.markRegionReplacedExistingAndCheckNeedsToBeRestarted(newRegion)) {
+            LOG.debug(
+              "Attempted to send a coprocessor service RPC to region {} which no"
+                + " longer exists, will attempt to send RPCs to the region(s) that replaced it",
+              region.getEncodedName());
+            restartCoprocessorServiceForRange(stubMaker, callable, callback, region.getStartKey(),
+              region.getEndKey(), newRegion, progress, span);
+          }
+          progress.markRegionFinished(region);
         } else {
-          callback.onRegionComplete(region, r);
-        }
+          if (!claimedRegionAssignment.get()) {
+            if (!progress.tryClaimRegionAssignment(region)) {
```

Comment:
There are two paths we need to consider now.
 * Original call
 * Recursive call (i.e. when we restart the request on the failed range)

Both places attempt to claim an "assignment" for the region after they receive a response, before actually propagating the request to the client's `CoprocessorCallback`